src/phoenix/trace/dsl/query.py

```diff
@@ -0,0 +1,248 @@
+from collections import defaultdict
+from dataclasses import dataclass, field, fields, replace
+from functools import cached_property, partial
+from types import MappingProxyType
+from typing import Any, Callable, ClassVar, Dict, Iterable, Iterator, List, Mapping, Sequence, Tuple
+
+import pandas as pd
+
+from phoenix.trace.dsl import SpanFilter
+from phoenix.trace.schemas import ATTRIBUTE_PREFIX, CONTEXT_PREFIX, Span
+from phoenix.trace.semantic_conventions import RETRIEVAL_DOCUMENTS
+
+_SPAN_ID = "context.span_id"
+_PRESCRIBED_POSITION_PREFIXES = {
+    RETRIEVAL_DOCUMENTS: "document_",
+    ATTRIBUTE_PREFIX + RETRIEVAL_DOCUMENTS: "document_",
+}
+_ALIASES = {
+    "span_id": "context.span_id",
+    "trace_id": "context.trace_id",
+}
+
+
+def _unalias(key: str) -> str:
+    return _ALIASES.get(key, key)
+
+
+@dataclass(frozen=True)
+class Projection:
+    key: str = ""
+    value: Callable[[Span], Any] = field(init=False, repr=False)
+    span_fields: ClassVar[Tuple[str, ...]] = tuple(f.name for f in fields(Span))
+
+    def __bool__(self) -> bool:
+        return bool(self.key)
+
+    def __post_init__(self) -> None:
+        key = _unalias(self.key)
+        object.__setattr__(self, "key", key)
+        if key.startswith(CONTEXT_PREFIX):
+            key = key[len(CONTEXT_PREFIX) :]
+            value = partial(self._from_context, key=key)
+        elif key.startswith(ATTRIBUTE_PREFIX):
+            key = self.key[len(ATTRIBUTE_PREFIX) :]
+            value = partial(self._from_attributes, key=key)
+        elif key in self.span_fields:
+            value = partial(self._from_span, key=key)
+        else:
+            value = partial(self._from_attributes, key=key)
+        object.__setattr__(self, "value", value)
+
+    def __call__(self, span: Span) -> Any:
+        return self.value(span)
+
+    @staticmethod
+    def _from_attributes(span: Span, key: str) -> Any:
+        return span.attributes.get(key)
+
+    @staticmethod
+    def _from_context(span: Span, key: str) -> Any:
+        return getattr(span.context, key, None)
+
+    @staticmethod
+    def _from_span(span: Span, key: str) -> Any:
+        return getattr(span, key, None)
+
+
+@dataclass(frozen=True)
+class Explosion(Projection):
```

Comment:
Don't forget to write down you ideas in docstrings.

Reply:
yup, I will do a fast-follow.